Fixed #24529 -- Allow double squashing of migrations

georgi-yanchev opened the pull request:
#### Trac ticket number
<!-- Replace XXXXX with the corresponding Trac ticket number, or delete the line and write "N/A" if this is a trivial PR. -->

ticket-24529

#### Branch description
This is a continuation of https://github.com/django/django/pull/14380 by @rtpg, which was closed due to inactivity. I have added a commit to address the [last comment](https://github.com/django/django/pull/14380#issuecomment-1127305761) in the pull request.

**Original Description:**
In order to support multi-level squashing, we need to be a bit smarter about how we traverse replacements. The solution here introduces some extra checks on squashed migrations (mainly a lookup for whether its replacements need to be squashed first), but the performance hit shouldn't be very large.

#### Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [x] I have attached screenshots in both light and dark modes for any UI changes.

jacobtylerwalls commented on docs/releases/4.1.txt:
```diff
@@ -329,6 +330,8 @@ Migrations
   ``AlterIndexTogether`` and ``AddIndex``, when moving indexes defined in the
   ``Meta.index_together`` to the
   :attr:`Meta.indexes <django.db.models.Options.indexes>`.
+* Squashed migrations can now themselves be squashed before being transitioned to
+  normal migrations.
```
Thanks for picking up the baton, just strolled by to remind you to move this release note to 5.2.txt.

georgi-yanchev replied:
Thanks for noticing, moved.

rtpg commented on django/core/management/commands/squashmigrations.py:
```diff
@@ -75,7 +74,9 @@ def handle(self, **options):
-        loader = MigrationLoader(connections[DEFAULT_DB_ALIAS])
+        loader = MigrationLoader(
+            connections[DEFAULT_DB_ALIAS], replace_migrations=False
+        )
```
I don't believe this is semantically what I wanted with the double squashing. If I squash A, B, and C into D, then D will have optimized migrations (I usually hand-optimize them too). Buy not using the operations from the squashed migration D, I'll lose all the work done.

This is the equivalent of having to re-resolve a merge conflict over and over again. Maybe fine in the end for most people? But not what I would be looking for as a result.

(Thank you for picking this up by the way! I had started working on this when I worked on a project with messy migrations but am no longer on that project...)

georgi-yanchev replied:
I see your concern. I've changed it to first try using the squashed migrations. If that fails then it will recreate the migration graph with all squashed migrations removed (replaced by original migrations) and try again.

georgi-yanchev commented:
Is there anything I can do to get it reviewed?

sarahboyce reviewed (commented):
Thank you @georgi-yanchev 

I downloaded Mariusz's test project and I still get the same error he reported [in his comment](https://github.com/django/django/pull/14380#issuecomment-1127305761)

georgi-yanchev commented:
Hi @sarahboyce. Thank you for taking the time to review this PR. I apologize that I couldn't find time to address your comments earlier.

> I downloaded Mariusz's test project and I still get the same error he reported https://github.com/django/django/pull/14380#issuecomment-1127305761

Looks like the problem is that `squashmigrations` command behaves differently, depending on which migrations have been already applied on the system where the command is run. Now I'm passing `connection=None` to the `MigrationLoader` constructor, so that the applied status is ignored. I added a test case with the migrations from Mariusz's test project.

georgi-yanchev commented on docs/topics/migrations.txt:
```diff
@@ -742,8 +742,11 @@ You must then transition the squashed migration to a normal migration by:
+   migrations, which might be useful for situations where every environment has
+   not yet run the original squashed migration set. But in general it is better
+   to transition squashed migrations to normal migrations to be able to clean up
+   older migration files.
```
Added.

georgi-yanchev commented on django/db/migrations/loader.py:
```diff
@@ -249,12 +250,73 @@ def build_graph(self):
+                if is_calculating:
+                    raise CommandError(
+                        f"Migration replacement loop found starting at {target}"
+                    )
```
I've removed the function altogether. It was called to check the applied status of 1st level replaced migrations. In fact we need to find the recursive list of replaced migrations, and then the old check is sufficient. 

The reason why we need the recursive list lies in Mariusz's [example](https://github.com/django/django/pull/14380#issuecomment-1127305761) Once we generate the 2nd level squash migration and try to migrate, we can apply the squashed only if all all or none of the migrations from the recursive list were applied.

georgi-yanchev commented on django/core/management/commands/squashmigrations.py:
```diff
@@ -75,7 +74,7 @@ def handle(self, **options):
         # Load the current graph state, check the app and migration they asked
         # for exists.
-        loader = MigrationLoader(connections[DEFAULT_DB_ALIAS])
+        loader = MigrationLoader(None)
```
connection is used to check which migrations have been applied on the system that runs the command. But we don't want generation of the squash migration to depend on the applied status.

georgi-yanchev commented:
Is there any chance we can get it merged into 5.2?

sarahboyce commented:
> Is there any chance we can get it merged into 5.2?

No I don't think we'll manage it, but will hopefully get it in 6.0

sarahboyce commented on django/db/migrations/loader.py:
```diff
@@ -250,12 +251,53 @@ def build_graph(self):
+            }
+
+            # this helper function handles the actual replacement
+            def do_replacement(key):
```
I think we don't need to define the helper in the method and can pull it out to be class level
```diff
--- a/django/db/migrations/loader.py
+++ b/django/db/migrations/loader.py
@@ -55,6 +55,7 @@ class MigrationLoader:
         self.applied_migrations = None
         self.ignore_no_migrations = ignore_no_migrations
         self.replace_migrations = replace_migrations
+        self.migrations_replaced = None
         if load:
             self.build_graph()
 
@@ -220,6 +221,58 @@ class MigrationLoader:
             if child is not None:
                 self.graph.add_dependency(migration, child, key, skip_validation=True)
 
+    def replace_migration(self, migration_key):
+        if self.migrations_replaced[migration_key]["finished_replacement"]:
+            return
+        if self.migrations_replaced[migration_key]["visited"]:
+            # This node is visited but have not finished the replacement, this means
+            # we have a circular dependency.
+            raise CommandError(
+                f"Cyclical squash replacement found, starting at {migration_key}"
+            )
+        self.migrations_replaced[migration_key]["visited"] = True
+        migration = self.migrations_replaced[migration_key]["migration"]
+        # Process potential squashed migrations that the migration replaces.
+        for replace_migration_key in migration.replaces:
+            if replace_migration_key in self.migrations_replaced:
+                self.replace_migration(replace_migration_key)
+
+        # Get replacement targets recursively.
+        previous_replaced_keys = migration.replaces
+        while True:
+            replaced_keys = []
+            for replaced_key in previous_replaced_keys:
+                replaced_migration = self.disk_migrations.get(replaced_key)
+                if replaced_migration and replaced_migration.replaces:
+                    replaced_keys.extend(replaced_migration.replaces)
+                else:
+                    replaced_keys.append(replaced_key)
+            if set(replaced_keys) == set(previous_replaced_keys):
+                break
+            previous_replaced_keys = replaced_keys
+        # Get applied status of each found replacement target.
+        applied_statuses = [
+            (target in self.applied_migrations) for target in replaced_keys
+        ]
+
+        # The replacing migration is only marked as applied if all of its
+        # replacement targets are applied.
+        if all(applied_statuses):
+            self.applied_migrations[migration_key] = migration
+        else:
+            self.applied_migrations.pop(migration_key, None)
+        # A replacing migration can be used if either all or none of its
+        # replacement targets have been applied.
+        if all(applied_statuses) or (not any(applied_statuses)):
+            self.graph.remove_replaced_nodes(migration_key, migration.replaces)
+        else:
+            # This replacing migration cannot be used because it is
+            # partially applied. Remove it from the graph and remap
+            # dependencies to it (#25945).
+            self.graph.remove_replacement_node(migration_key, migration.replaces)
+
+        self.migrations_replaced[migration_key]["finished_replacement"] = True
+
     def build_graph(self):
         """
         Build a migration dependency graph using both the disk and database.
@@ -251,73 +304,20 @@ class MigrationLoader:
             self.add_external_dependencies(key, migration)
         # Carry out replacements where possible and if enabled.
         if self.replace_migrations:
-            # migration_work[key] = [migration, visited, finished_replacement]
-            # this is mainly used to track cyclical dependencies and which
-            # migrations we have finished replacing, and is needed because
-            # squashed migrations can depend on other squashed migrations
-            migration_work = {
-                key: [migration, False, False]
+            # Use to track cyclical dependencies and which migrations we have
+            # finished replacing, as squashed migrations can depend on other
+            # squashed migrations.
+            self.migrations_replaced = {
+                key: {
+                    "migration": migration,
+                    "visited": False,
+                    "finished_replacement": False,
+                }
                 for key, migration in self.replacements.items()
             }
 
-            # this helper function handles the actual replacement
-            def do_replacement(key):
-                migration, visited, finished_replacement = migration_work[key]
-                if finished_replacement:
-                    # we have already replaced this item
-                    return
-                if visited:
-                    # we visited this node but have not finished the replacement
-                    # this means we have a circular dependency
-                    raise CommandError(
-                        f"Cyclical squash replacement found, starting at {key}"
-                    )
-                # begin visiting this one
-                migration_work[key][1] = True
-                # before processing this replacement, process
-                # potential squashed elements in its replaces
-                for key2 in migration.replaces:
-                    if key2 in migration_work:
-                        do_replacement(key2)
-
-                # Get replacement targets recursively.
-                previous_replaced_keys = migration.replaces
-                while True:
-                    replaced_keys = []
-                    for replaced_key in previous_replaced_keys:
-                        replaced_migration = self.disk_migrations.get(replaced_key)
-                        if replaced_migration and replaced_migration.replaces:
-                            replaced_keys.extend(replaced_migration.replaces)
-                        else:
-                            replaced_keys.append(replaced_key)
-                    if set(replaced_keys) == set(previous_replaced_keys):
-                        break
-                    previous_replaced_keys = replaced_keys
-                # Get applied status of each found replacement target.
-                applied_statuses = [
-                    (target in self.applied_migrations) for target in replaced_keys
-                ]
-
-                # The replacing migration is only marked as applied if all of
-                # its replacement targets are.
-                if all(applied_statuses):
-                    self.applied_migrations[key] = migration
-                else:
-                    self.applied_migrations.pop(key, None)
-                # A replacing migration can be used if either all or none of
-                # its replacement targets have been applied.
-                if all(applied_statuses) or (not any(applied_statuses)):
-                    self.graph.remove_replaced_nodes(key, migration.replaces)
-                else:
-                    # This replacing migration cannot be used because it is
-                    # partially applied. Remove it from the graph and remap
-                    # dependencies to it (#25945).
-                    self.graph.remove_replacement_node(key, migration.replaces)
-                # at the end, mark this node as having been visited completely
-                migration_work[key][2] = True
-
-            for key in self.replacements.keys():
-                do_replacement(key)
+            for migration_key in self.replacements.keys():
+                self.replace_migration(migration_key)
         # Ensure the graph is consistent.
         try:
```

sarahboyce commented on tests/migrations/test_commands.py:
```diff
@@ -2939,6 +2942,148 @@ def test_squashmigrations_squashes(self):
             "  you can delete them.\n" % squashed_migration_file,
         )
 
+    def test_squashmigrations_replaces_loop(self):
+        """
+        our migration loader will properly detect if there is a loop inside of our
+        squashing code, and will raise an error in that case.
+        """
+        out = io.StringIO()
+        with self.temporary_migration_module(
+            module="migrations.test_migration_squash_loop"
+        ):
+            # we expect to hit a squash replacement cycle check error, but the actual
+            # failure is dependent on the order in which the files are read on disk.
```
```suggestion
    def test_squashmigrations_replacement_cycle(self):
        out = io.StringIO()
        with self.temporary_migration_module(
            module="migrations.test_migration_squash_loop"
        ):
            # Hits a squash replacement cycle check error, but the actual failure is
            # dependent on the order in which the files are read on disk.
```
(nit)

sarahboyce commented on tests/migrations/test_commands.py:
```diff
@@ -2939,6 +2942,148 @@ def test_squashmigrations_squashes(self):
+                interactive=False,
+            )
+
+            # confirm the replaces mechanism holds the squashed migration
+            # (and not what it squashes, as the squash operations are what
+            #  end up being used)
+
+            loader = MigrationLoader(connection)
+            migration = loader.disk_migrations[("migrations", "0001_double_squash")]
```
```suggestion
            loader = MigrationLoader(connection)
            migration = loader.disk_migrations[("migrations", "0001_double_squash")]
            # Confirm the replaces mechanism holds the squashed migration
            # (and not what it squashes, as the squash operations are what
            # end up being used).
```

sarahboyce commented on tests/migrations/test_commands.py:
```diff
@@ -2939,6 +2942,148 @@ def test_squashmigrations_squashes(self):
+                ],
+            )
+
+            # check the migrations that would get run
+            # check that we have migrations.0001_double_squash
+            # migrations.6_auto, and migrations.7_auto
+
```
```suggestion
```

sarahboyce commented on tests/migrations/test_commands.py:
```diff
@@ -2939,6 +2942,148 @@ def test_squashmigrations_squashes(self):
+        with self.temporary_migration_module(
+            module="migrations.test_migrations_squash_partially_applied"
+        ):
+            # Apply first 2 migrations
```
```suggestion
            # Apply first 2 migrations.
```

sarahboyce reviewed (commented):
Thank you for the fix @georgi-yanchev :partying_face: 

I have added a few stylist comments.
Can you squash to a sinlge commit, rebase on main, and move the release note to 6.0 (and update `.. versionchanged:: 5.2` to `.. versionchanged:: 6.0`)? :star:

georgi-yanchev commented:
Thanks @sarahboyce.

I applied your suggested changes, moved the documentation to 6.0 and squashed everything into a single commit on top of the latest `main`. 

I see 6 test are failing. 4 of them because the disk is full on the test worker and the other 2 because a unrelated XML file is empty??? Also I see that the same tests failed in the same way for other pull requests, so I assume they are just flaky?

Can we merge the PR now, or should we re-run these tests once the flakiness is resolved?

sarahboyce commented on django/db/migrations/loader.py:
```diff
@@ -219,6 +221,58 @@ def add_external_dependencies(self, key, migration):
+            if replace_migration_key in self.migrations_replaced:
+                self.replace_migration(replace_migration_key)
+
+        # Get replacement targets recursively.
+        previous_replaced_keys = migration.replaces
+        while True:
+            replaced_keys = []
+            for replaced_key in previous_replaced_keys:
+                replaced_migration = self.disk_migrations.get(replaced_key)
+                if replaced_migration and replaced_migration.replaces:
+                    replaced_keys.extend(replaced_migration.replaces)
+                else:
+                    replaced_keys.append(replaced_key)
+            if set(replaced_keys) == set(previous_replaced_keys):
+                break
+            previous_replaced_keys = replaced_keys
```
I think I will pull this into a helper:
```diff
--- a/django/db/migrations/loader.py
+++ b/django/db/migrations/loader.py
@@ -221,6 +221,17 @@ class MigrationLoader:
             if child is not None:
                 self.graph.add_dependency(migration, child, key, skip_validation=True)
 
+    def _resolve_replaced_migration_keys(self, migration):
+        resolved_keys = set()
+        for migration_key in set(migration.replaces):
+            migration_entry = self.disk_migrations.get(migration_key)
+            if migration_entry and migration_entry.replaces:
+                replace_keys = self._resolve_replaced_migration_keys(migration_entry)
+                resolved_keys.update(replace_keys)
+            else:
+                resolved_keys.add(migration_key)
+        return resolved_keys
+
     def replace_migration(self, migration_key):
         if self.migrations_replaced[migration_key]["finished_replacement"]:
             return
@@ -237,19 +248,7 @@ class MigrationLoader:
             if replace_migration_key in self.migrations_replaced:
                 self.replace_migration(replace_migration_key)
 
-        # Get replacement targets recursively.
-        previous_replaced_keys = migration.replaces
-        while True:
-            replaced_keys = []
-            for replaced_key in previous_replaced_keys:
-                replaced_migration = self.disk_migrations.get(replaced_key)
-                if replaced_migration and replaced_migration.replaces:
-                    replaced_keys.extend(replaced_migration.replaces)
-                else:
-                    replaced_keys.append(replaced_key)
-            if set(replaced_keys) == set(previous_replaced_keys):
-                break
-            previous_replaced_keys = replaced_keys
+        replaced_keys = self._resolve_replaced_migration_keys(migration)
         # Get applied status of each found replacement target.
```

georgi-yanchev replied:
Thanks, I applied the patch, squashed and force-pushed again.

georgi-yanchev commented:
> @georgi-yanchev can you grant me permissions to push to your repo?

No, unfortunately. The github account where the repo is hosted is owned by my company and granting access to external people is tricky. If you want to push more changes, then I can clone `django` into my personal account (where I can give you access) and then I guess I'll have to open a new pull request.

sarahboyce reviewed (approved):
Thank you for the updates!

I needed to push some tweaks (including minor things like make the commit past tense, align the naming of the folders of test migrations etc)
So created a PR rather than comment each nit pick: https://github.com/django/django/pull/19082
